File: chrome/browser/ui/android/extensions/extension_action_context_menu_bridge.cc
// Copyright 2025 The Chromium Authors
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "chrome/browser/ui/android/extensions/extension_action_context_menu_bridge.h"

#include "base/android/jni_android.h"
#include "base/android/jni_string.h"
#include "chrome/browser/extensions/extension_context_menu_model.h"
#include "chrome/browser/profiles/profile.h"
#include "chrome/browser/ui/browser_window/public/browser_window_interface.h"
#include "chrome/browser/ui/toolbar/toolbar_actions_model.h"
#include "content/public/browser/web_contents.h"
#include "extensions/browser/extension_registry.h"

// Must come after all headers that specialize FromJniType() / ToJniType().
#include "chrome/browser/ui/android/extensions/jni_headers/ExtensionActionContextMenuBridge_jni.h"

using base::android::JavaParamRef;
using base::android::ScopedJavaLocalRef;
using content::WebContents;
using extensions::Extension;
using extensions::ExtensionContextMenuModel;
using extensions::ExtensionRegistry;

namespace extensions {

ExtensionActionContextMenuBridge::ExtensionActionContextMenuBridge(
    BrowserWindowInterface* browser,
    const ToolbarActionsModel::ActionId& action_id,
    content::WebContents* web_contents,
    ExtensionContextMenuModel::ContextMenuSource context_menu_source) {
  Profile* profile = browser->GetProfile();
  auto* registry = extensions::ExtensionRegistry::Get(profile);
  scoped_refptr<const extensions::Extension> extension =
      registry->enabled_extensions().GetByID(action_id);
  DCHECK(extension);

  bool is_pinned =
      ToolbarActionsModel::Get(profile)->IsActionPinned(extension->id());

  extension_context_menu_model_ = std::make_unique<ExtensionContextMenuModel>(
      extension.get(), browser, is_pinned, /* delegate= */ nullptr,
      /* can_show_icon_in_toolbar= */ true, context_menu_source);

  menu_model_bridge_ = std::make_unique<ui::MenuModelBridge>(
      extension_context_menu_model_->AsWeakPtr());
}

ExtensionActionContextMenuBridge::~ExtensionActionContextMenuBridge() {}

ScopedJavaLocalRef<jobject>
ExtensionActionContextMenuBridge::GetMenuModelBridge(JNIEnv* env) {
  return menu_model_bridge_->GetJavaObject().AsLocalRef(env);
}

void ExtensionActionContextMenuBridge::Destroy(JNIEnv* env) {
  delete this;
}

static jlong JNI_ExtensionActionContextMenuBridge_Init(
    JNIEnv* env,
    jlong browser_window_interface_ptr,
    ToolbarActionsModel::ActionId& action_id,
    content::WebContents* web_contents,
    ExtensionContextMenuModel::ContextMenuSource context_menu_source) {
  BrowserWindowInterface* browser =
      reinterpret_cast<BrowserWindowInterface*>(browser_window_interface_ptr);
  auto* bridge = new ExtensionActionContextMenuBridge(
      browser, action_id, web_contents, context_menu_source);
  // The bridge is owned by the Java object and will be destroyed when
  // ExtensionActionContextMenuBridge.destroy() is called.
  return reinterpret_cast<jlong>(bridge);
}

}  // namespace extensions

DEFINE_JNI(ExtensionActionContextMenuBridge)

File: chrome/browser/ui/android/extensions/java/src/org/chromium/chrome/browser/ui/extensions/ExtensionActionContextMenuBridge.java
// Copyright 2025 The Chromium Authors
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

package org.chromium.chrome.browser.ui.extensions;

import org.jni_zero.JNINamespace;
import org.jni_zero.JniType;
import org.jni_zero.NativeMethods;

import org.chromium.base.lifetime.Destroyable;
import org.chromium.base.lifetime.LifetimeAssert;
import org.chromium.build.annotations.NullMarked;
import org.chromium.build.annotations.Nullable;
import org.chromium.chrome.browser.extensions.ContextMenuSource;
import org.chromium.chrome.browser.ui.browser_window.ChromeAndroidTask;
import org.chromium.content_public.browser.WebContents;
import org.chromium.ui.listmenu.MenuModelBridge;
import org.chromium.ui.modelutil.MVCListAdapter.ModelList;

/**
 * This class owns the model that hosts the contents of toolbar action context menu.
 *
 * <p>This class implements {@link Destroyable} to ensure proper cleanup of its associated native
 * resources. The native counterpart's lifetime is tied to this Java object; when {@link #destroy()}
 * is called, the native object is also destroyed.
 */
@NullMarked
@JNINamespace("extensions")
public class ExtensionActionContextMenuBridge implements Destroyable {

    private long mNativeExtensionActionContextMenuBridge;

    private final @Nullable LifetimeAssert mLifetimeAssert = LifetimeAssert.create(this);

    public ExtensionActionContextMenuBridge(
            ChromeAndroidTask task,
            String actionId,
            WebContents webContents,
            @ContextMenuSource int contextMenuSource) {
        mNativeExtensionActionContextMenuBridge =
                ExtensionActionContextMenuBridgeJni.get()
                        .init(
                                task.getOrCreateNativeBrowserWindowPtr(),
                                actionId,
                                webContents,
                                contextMenuSource);
    }

    /** Returns the {@link ModelList} of the content of the toolbar action's context menu. */
    public ModelList getModelList() {
        MenuModelBridge menuModelBridge =
                ExtensionActionContextMenuBridgeJni.get()
                        .getMenuModelBridge(mNativeExtensionActionContextMenuBridge);
        return menuModelBridge.populateModelList();
    }

    /**
     * Cleans up the resources associated with this popup.
     *
     * <p>This method must be called when the {@link MenuModel} is no longer needed.
     */
    @Override
    public void destroy() {
        if (mNativeExtensionActionContextMenuBridge != 0) {
            ExtensionActionContextMenuBridgeJni.get()
                    .destroy(mNativeExtensionActionContextMenuBridge);
            mNativeExtensionActionContextMenuBridge = 0;
            LifetimeAssert.destroy(mLifetimeAssert);
        }
    }

    @NativeMethods
    public interface Native {
        long init(
                long androidBrowserWindowPtr,
                @JniType("std::string") String actionId,
                @JniType("content::WebContents*") WebContents webContents,
                @JniType("ExtensionContextMenuModel::ContextMenuSource") int contextMenuSource);

        MenuModelBridge getMenuModelBridge(long nativeExtensionActionContextMenuBridge);

        void destroy(long nativeExtensionActionContextMenuBridge);
    }
}
